bolt-boost/src/server.rs

```diff
@@ -0,0 +1,582 @@
+use alloy::{
+    primitives::{utils::format_ether, B256, U256},
+    rpc::types::beacon::{relay::ValidatorRegistration, BlsPublicKey},
+};
+use async_trait::async_trait;
+use axum::{
+    extract::{Path, State},
+    http::{header::USER_AGENT, HeaderMap, HeaderValue, StatusCode},
+    response::IntoResponse,
+    routing::{get, post},
+    Json, Router,
+};
+use cb_common::{
+    config::PbsConfig,
+    constants::APPLICATION_BUILDER_DOMAIN,
+    pbs::{
+        error::{PbsError, ValidationError},
+        GetHeaderResponse, RelayClient, SignedExecutionPayloadHeader, EMPTY_TX_ROOT_HASH,
+        HEADER_SLOT_UUID_KEY, HEADER_START_TIME_UNIX_MS,
+    },
+    signature::verify_signed_message,
+    types::Chain,
+    utils::{get_user_agent_with_version, ms_into_slot},
+};
+use commit_boost::prelude::*;
+use eyre::Result;
+use futures::{future::join_all, stream::FuturesUnordered, StreamExt};
+use serde::Serialize;
+use std::{
+    collections::HashMap,
+    time::{Duration, Instant},
+};
+use tokio::time::sleep;
+use tracing::{debug, error, info, warn, Instrument};
+
+use super::{
+    constraints::ConstraintsCache,
+    error::PbsClientError,
+    proofs::verify_multiproofs,
+    types::{
+        Config, GetHeaderParams, GetHeaderWithProofsResponse, RequestConfig, SignedConstraints,
+        SignedDelegation, SignedExecutionPayloadHeaderWithProofs, SignedRevocation,
+    },
+};
+
+const SUBMIT_CONSTRAINTS_PATH: &str = "/constraints/v1/builder/constraints";
+const DELEGATE_PATH: &str = "/constraints/v1/builder/delegate";
+const REVOKE_PATH: &str = "/constraints/v1/builder/revoke";
+const GET_HEADER_WITH_PROOFS_PATH: &str =
+    "/eth/v1/builder/header_with_proofs/:slot/:parent_hash/:pubkey";
+
+const TIMEOUT_ERROR_CODE: u16 = 555;
+
+// lazy_static! {
+//     pub static ref CHECK_RECEIVED_COUNTER: IntCounter =
+//         IntCounter::new("checks", "successful /check requests received").unwrap();
+// }
+
+// Extra state available at runtime
+#[derive(Clone)]
+pub struct BuilderState {
+    #[allow(unused)]
+    config: Config,
+    constraints: ConstraintsCache,
+}
+
+impl BuilderApiState for BuilderState {}
+
+impl BuilderState {
+    pub fn from_config(config: Config) -> Self {
+        Self {
+            config,
+            constraints: ConstraintsCache::new(),
+        }
+    }
+}
+
+/// An extended builder-API that implements the constraints-API as defined in
+/// the spec: <https://chainbound.github.io/bolt-docs/api/builder>.
+///
+/// The added endpoints are defined in [extra_routes](ConstraintsApi::extra_routes).
+pub struct ConstraintsApi;
+
+#[async_trait]
+impl BuilderApi<BuilderState> for ConstraintsApi {
+    /// Register a validator with the builder.
+    ///
+    /// We intercept this call since it happens periodically and we use it to clean
+    /// up old constraints.
```

Comment:
I get why this happens, but what happens if the validator doesn't call this endpoint regularly? we risk going OOM?

Reply:
Every beacon node registers their validators every epoch, so seems like a pretty big edge case

Reply:
Great, works for me then!